### PR TITLE
Various API additions to support filter development

### PR DESCRIPTION
**Pull request by imikejackson**
* CropImageGeometry: Parallelize the array copying
* Segment Features: Add ability to cancel the segmentation loop. Clean up test code
* DataArrayUtilities: Added functions to support using Bool or UInt8 for Mask Arrays
* Create base class for AlignSections Algorithm.

### CODE REVIEW

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]
```cpp
/**
 * @brief TaskRunner runs a functor in a tbb::task_group if multicore is enabled.
 * It does not spawn more tasks than there are threads. Once the max thread count
 * is reached, the task group waits until all current tasks have been completed.
 * If multicore is not enabled, functors run sequentially. wait() must be called
 * before the object destructs.
 */
class TaskRunner
{
public:
  TaskRunner()
  : m_MaxThreadCount(std::thread::hardware_concurrency()) // std::thread::hardware_concurrency() returns ZERO if not defined on this platform
  {
  }

  TaskRunner(uint32 maxThreads)
  : m_MaxThreadCount(maxThreads)
  {
  }

  TaskRunner(const TaskRunner&) = delete;
  TaskRunner(TaskRunner&&) = delete;

  TaskRunner& operator=(const TaskRunner&) = delete;
  TaskRunner& operator=(TaskRunner&&) = delete;

  ~TaskRunner() = default;

  auto wait()
  {
#ifdef COMPLEX_ENABLE_MULTICORE
    return m_TaskGroup.wait();
#endif
  }

  template <class FunctorT>
  void run(FunctorT&& func)
  {
#ifdef COMPLEX_ENABLE_MULTICORE
    if(m_CurrentThreadCount == m_MaxThreadCount)
    {
      m_TaskGroup.wait();
      m_CurrentThreadCount = 0;
    }
    m_TaskGroup.run(std::forward<FunctorT>(func));
    m_CurrentThreadCount++;
#else
    func();
#endif
  }

private:
#ifdef COMPLEX_ENABLE_MULTICORE
  tbb::task_group m_TaskGroup;
  uint32 m_CurrentThreadCount = 0;
  uint32 m_MaxThreadCount = 1;
#endif
};
```
This should work for `CropImageGeometry` and `AlignSections`.

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/CropImageGeometry.cpp [hunk 1]
```diff
@@ -1,24 +1,27 @@
...
+#include "complex/Utilities/ParallelData3DAlgorithm.hpp"
+#include "complex/Utilities/ParallelDataAlgorithm.hpp"
 #include "complex/Utilities/SamplingUtils.hpp"
 
+#ifdef COMPLEX_ENABLE_MULTICORE
+#define RUN_TASK g->run
+#else
+#define RUN_TASK
+#endif
+
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/CropImageGeometry.cpp:24]:
```suggestion
```

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

File: src/complex/Utilities/DataArrayUtilities.cpp [hunk 2]
```diff
@@ -208,4 +208,31 @@ Result<> ResizeAndReplaceDataArray(DataStructure& dataStructure, const DataPath&
 
   return MakeErrorResult(-401, fmt::format("The input array at DataPath '{}' was of an unsupported type", dataPath.toString()));
 }
+
+std::unique_ptr<MaskCompare> InstantiateMaskCompare(const DataStructure& dataStructure, const DataPath& maskArrayPath)
+{
+  if(maskArrayPath.empty())
+  {
+    return {};
+  }
```
[the remark this answers is not in the document]
Comment (imikejackson) [reply src/complex/Utilities/DataArrayUtilities.cpp:217]: If that check isn't in there then the next line will throw an exception if the path is empty.  Thoughts?

[the remark this answers is not in the document]
Comment (imikejackson) [reply src/complex/Utilities/DataArrayUtilities.cpp]: I ended up putting a try.. catch block in the calling code instead.

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/CropImageGeometry.cpp (hunk 1 above)
Comment (imikejackson) [reply src/Plugins/ComplexCore/src/ComplexCore/Filters/CropImageGeometry.cpp]: Done in latest commit

**Review by JDuffeyBQ — APPROVED** [review]

**Pull request merged**